thrift/src/main/java/com/linecorp/armeria/client/thrift/THttpClientInvocationHandler.java

```diff
@@ -119,31 +124,51 @@ private Object invokeClientMethod(Method method, @Nullable Object[] args) throws
             }
         }
 
-        try {
-            final RpcResponse reply;
-            if (fragment != null) {
-                reply = delegate().executeMultiplexed(
-                        path, params.clientType(), fragment, method.getName(), args);
-            } else {
-                reply = delegate().execute(path, params.clientType(), method.getName(), args);
-            }
+        try (ClientRequestContextCaptor captor = Clients.newContextCaptor()) {
+            try {
+                final RpcResponse reply;
+                if (fragment != null) {
+                    reply = delegate().executeMultiplexed(
+                            path, params.clientType(), fragment, method.getName(), args);
+                } else {
+                    reply = delegate().execute(path, params.clientType(), method.getName(), args);
+                }
 
-            if (callback != null) {
-                AsyncMethodCallbacks.transfer(reply, callback);
-                return null;
-            } else {
-                try {
-                    return reply.get();
-                } catch (ExecutionException e) {
-                    throw Exceptions.peel(e);
+                final ClientRequestContext ctx = captor.get();
+                if (callback != null) {
+                    reply.handle(ctx.makeContextAware((res, cause) -> {
```

Comment:
I think we can directly mount the context here and below instead of using `makeContextAware`.

Reply:
https://github.com/line/armeria/pull/2383/files#diff-6ab8399d40ca91dd25785dc822606306R167

This line should be called by the same executor so I chose this way which is let the thread which complete the `reply` also does the jobs in callback.
But I'm not so strong this and I can put the job in the executor.

Reply:
I'm not sure I understand, probably since I wasn't clear either - I didn't mean use the executor, but just something like this to directly mount the context without creating a wrapped callable

```java
reply.handle((res, cause) -> {
  try (SafeCloseable unused = ctx.push()) {
    if (cause != null) {
                     invokeOnError(callback, cause);
                } else {
                    callback.onComplete(res);
                            }
                        } catch (Exception e) {
                            CompletionActions.log(e);
                        }
                        return null;
                    }));
```

Reply:
Well actually, I can't get an `executor` in that case. Is it okay to execute the callback`(invokeOnError(callback, cause);)` with the different thread in this case?

Reply:
By the way, as far as the thread, from what I understand, the thread running the `handle` callback will be the thread in `delegate().execute(` and isn't related to this code directly. Though I guess this code, `delegate().execute` will in practice be the same thread, which happens to be `ctx.contextAwareExecutor` too. JFYI, I don't think it affects how we mount the context here.

Reply:
Well, it could be the thread which calls `reply.handle()` (i,e, reply is completed when attaching callback. 😭  This is why previous [PR](https://github.com/line/armeria/commit/845f482e6c39e1c55b4434473466ffd7af4409d5#commitcomment-36690155) was not working )

Reply:
> I'm not sure I understand, probably since I wasn't clear either - I didn't mean use the executor, but just something like this to directly mount the context without creating a wrapped callable

Yeah, I can do that and I misunderstood it. 😅

Reply:
Aha - thanks I understand the threading issue now. I never realized the non-async callback methods had such a dangerous caveat to them